search tags in reverse order

arsenalzp opened the pull request:
This PR is intended to implement feature requested in issue #26053 

As tags list might be either in ascending or descending order I decided to use boolean option `--reverse-order` instead of `--order=newest` and `--order=older`. 

The option `--reverse-order` is meaningful by itself.
<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?
<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".
None
For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
None
```

arsenalzp commented:
> Bool options don't scale well, what when the next person likes it sorted in a different way? I think sticking with --sort= is best because that also aligns with other commands such as `podman ps`.
> 
> Also looking at the code I don't think any sorting is guaranteed whatsoever currently. So the assumption that it is sorted today is somewhat wrong. The code returns the results in the order the registry returned it AFAICT so it relies on the implemenation details of each registry and I doubt the image spec would have that defined in any way. @mtrmac might know more abut that
> 
> So in short if we do like to have the option to sort then we must have the option to sort in both orders

Hello,
I didn't know search return is non-Idempotent and depends on registries implementation.
So, two way sort does make sense here. I will rewrite code and related test.

mtrmac commented:
First, let’s talk about how this does not implement #26053 : that asks for “newest” tags, not “alphabetically latest” tags.
- I _suspect_ the issue asked for “tagged latest”, i.e. “sort by time, return the most recently tagged one”. That’s outright impossible, “when was the tag added” is not a part of the registry API. (I suppose it would be possible to inspect each individual image and try to figure out when it was _built_, but that’s not the same thing.
- Alternatively, the issue might have meant “numerically latest version”, and this clearly does not do that, because `9` < `10`, but `"10"` < `"9"`. The _definition_ of “latest version” is unclear, because there is no universal standard for version number formatting, and the tag has no format defined (so it _might_ be SemVer but pretty frequently is not).

But, the most important part is that the whole _idea_ of #26053 is badly targeted: If the goal is to list the latest tag, that should use a “list tags in a repo” API, not an ill-defined “registry search with no defined semantics” API. Registries can, and do, limit the number of results. So, the latest tag might not be in the list at all!

My first recommendation is to use `skopeo list-tags` in a pipeline with a _custom_ sorting command which precisely implements “latest” ordering _as defined for that repo_. Maybe we could do better, I’m not sure, but the whole requested use case of #26053 is clearly worse.

---

> The code returns the results in the order the registry returned

Yes: The order, and the _subset_ the registry chose to return.

arsenalzp commented:
> > But, the most important part is that the whole _idea_ of #26053 is badly targeted: If the goal is to list the latest tag, that should use a “list tags in a repo” API, not an ill-defined “registry search with no defined semantics” API. Registries can, and do, limit the number of results. So, the latest tag might not be in the list at all!
> 
> Oops, `podman search --list-tags` actually is not a _search_ at all, it _is_ the “list tags in a repo” call. … But it, by default, limits the returned tags to an _arbitrary_ subset of 25, and there seems to be no convenient way to return all of them (`--limit 99999999999` I suppose).

So, what it the decision for this particular issue? Should it be closed?

mtrmac commented:
I’m leaning towards closing (as long as piping the output of `Podman` to a separate sorting tool is convenient enough), but let’s see what others think.

Luap99 commented:
Maybe @justin-octo can clarify on the exact request based on the points above.

But yeah logically one can do `podman search --format {{.Tag}}  --list-tags $IMAGE` and the pipe that to whatever sorting command they like.

If the ask is just sorting by string asc/desc like this PR then I would not object to adding it but adding more complex stuff such as semver parsing seems out of scope to me.

justin-octo commented:
When I do:
```
podman search --format {{.Tag}} --list-tags docker.io/redhat/ubi9
```
I only get the first 26 Tags (oldest from what I can tell).  If I read above correctly, this order is determined by the registry.  So, obviously, if there are hundreds of tags, I'm going to have to "page" all the way to the end to find the newest tags.

This seems like a lot of wasted data if all I want is the 25 newest tags.  So, being able to reverse that sort order is what I was hoping for.  And get the 25 newest, instead of the 25 oldest (or list them all, and then deal with hundreds of lines and multiple pages).

mtrmac commented:
> So, the primary goal is to be able to easily identify the image tagged with the version number where the image is the same as the one tagged "latest".

That would be interesting but the vendor-neutral registry protocol has no support for looking up tags by images (by digests). The code would have to fetch the manifest for each tag, as one HTTP request per tag.

It’s software that could be written but it’s such a number of requests, especially in the worst cases, that I don’t think this should be available as a convenient option. Of course it can be scripted using `skopeo list-tags` and `skopeo inspect`, but I think we _need_ to steer users to different workflows.

Presumably the author of the image knows which image is the latest, and can provide that information using some other channel than a registry. Or maybe that specific registry implementation has its own private API that exposes something relevant.

> Being able to sort "pushed date" by most recent

That would also be interesting, and it’s, again, not something available in the vendor-neutral registry protocol.

kolyshkin reviewed (commented):
1. If the searchReport is already sorted, then you can just reverse it.
2. Your commit message says "remove duplicated import" and I'm not sure what this refers to.

arsenalzp commented:
> 1. If the searchReport is already sorted, then you can just reverse it.
Yes, I did reverse.

>     2. Your commit message says "remove duplicated import" and I'm not sure what this refers to.
I added another package import with different alias, I removed it.

By the way there is no decision if introduced changes are needed. I guess this PR will be closed.
